scipy/signal/_peak_finding_utils.pyx

```diff
@@ -74,3 +74,91 @@ def _argmaxima1d(np.float64_t[:] x not None):
 
     maxima.resize(m, refcheck=False)  # Keep only valid part of array memory.
     return maxima
+
+
+@cython.wraparound(False)
+@cython.boundscheck(False)
+def _peak_prominences(np.float64_t[:] x not None,
+                      np.intp_t[:] peaks not None,
+                      np.intp_t wlen):
+    """
+    Calculate the prominence of each peak in a signal.
+
+    Parameters
+    ----------
+    x : ndarray
+        A signal with peaks.
+    peaks : ndarray
+        Indices of peaks in `x`.
+    wlen : np.intp
+        A window length in samples (see `peak_prominences`). If -1 is given the
+        entire signal `x` is used.
+
+    Returns
+    -------
+    prominences : ndarray
+        The calculated prominences for each peak in `peaks`.
+    left_bases, right_bases : ndarray
+        The peaks' bases as indices in `x` to the left and right of each peak.
+
+    Notes
+    -----
+    This is the inner function to `peak_prominences`.
+
+    .. versionadded:: 1.1.0
+    """
+    cdef np.ndarray[np.float64_t, ndim=1] prominences
+    cdef np.ndarray[np.intp_t, ndim=1] left_bases, right_bases
+    cdef np.float64_t[:] window
+    cdef np.float64_t peak_height, left_min, right_min
+    cdef np.intp_t peak_nr, peak, wleft, wright, i, i_min, i_max
+
+    # Allocate returned objects
+    prominences = np.empty(peaks.shape[0], dtype=np.float64)
+    left_bases = np.empty(peaks.shape[0], dtype=np.intp)
+    right_bases = np.empty(peaks.shape[0], dtype=np.intp)
```

Comment:
Alternately, use `np.empty_like`

Reply:
I thought about it but I'm not sure whether

* it would actually be faster
* I wouldn't have problems with the memory layout because by default `empty_like` tries to match the memory layout as close as possible... which I think would have been a problem if `x` was a non-contiguous array.

Reply:
Furthermore there is no measurable difference between using `peaks.shape[0]` repeatedly and storing it inside a typed variable. I find the first case more readable.

Reply:
Works for me.